lsp/server/src/server.ts

```diff
@@ -155,6 +155,56 @@ connection.languages.diagnostics.on(async (params) => {
     }
 });
 
+connection.onCodeAction((params) => {
+    const textDocument = documentCache.get(params.textDocument.uri);
+    const diagnostics = params.context.diagnostics;
+    if (textDocument === undefined || diagnostics.length === 0) {
+        return undefined;
+    }
+
+    const result: CodeAction[] = [];
+
+    diagnostics.forEach((diagnostic) => {
+        // generate the two suppressing quick fixes
+        const { data: producerId } = diagnostic;
```

Comment:
I still don't understand what this does. Why not use `diagnostic.producerId`?

Reply:
changed  `const { data: producerId } = diagnostic; ` to `const producerId = diagnostic.data;` 

About your question **Why not use diagnostic.producerId?**,  I tried, the producerId doesn't persist cross when it's being shuffled between server to client and client back to server.  Here's the exact steps I did:

1. extended Diagnostic interface to have new property 'produceId' using [Declaration Merging](https://www.typescriptlang.org/docs/handbook/declaration-merging.html)
2. set diagnostic.producerId when the diagnostic is generated for lsp client to show it. 
3. when mousing over the rendered diagnostic, Server.onCodeAction is called with a parameter of CodeActionParams.  The parameter holds an instance of the diagnostic. the producerId of that diagnostic is undefined.  

Seems that lsp system doesn't like the producerId and clean it away. 

That's my findings.